src/clr-angular/data/datagrid/providers/column-order-model.service.ts

```diff
@@ -0,0 +1,107 @@
+ * 2. Updating the flex order data and tells ColumnOrderCoordinatorService when it should broadcast
+ * 3. Returning models at the previous and next flex orders
+ */
+
+@Injectable()
+export class ColumnOrderModelService {
+  constructor(private columnOrderCoordinatorService: ColumnOrdersCoordinatorService, private domAdapter: DomAdapter) {}
+
+  public flexOrder: number;
+
+  public headerEl: any;
+
+  public hideableColumnModel: DatagridHideableColumnModel;
+
+  get columnGroupId() {
+    return this.columnOrderCoordinatorService.columnGroupId;
+  }
+
+  get isAtFirst(): boolean {
+    return this.flexOrder === 0;
+  }
+
+  get isAtEnd(): boolean {
+    return this.flexOrder === this.columnOrderCoordinatorService.orderModels.length - 1;
+  }
+
+  get isHidden(): boolean {
+    return this.hideableColumnModel && this.hideableColumnModel.hidden;
+  }
+
+  public dropReceived(dropData: any) {
+    // updates column orders
+    // broadcasts updated order changes
+    console.log(dropData);
+  }
+
+  get nextVisibleColumnModel(): ColumnOrderModelService {
+    // collect all columns that appears after this column
+    // that means collect all models with greater flex order value
+    const nextVisibleColumnModels = this.columnOrderCoordinatorService.orderModels.filter(
+      model => model.flexOrder > this.flexOrder && !model.isHidden
+    );
+
+    // current column could be the one that visually appears at the end;
+    if (nextVisibleColumnModels.length === 0) {
+      return;
+    }
+
+    // to find the next immediate visible columns model
+    // find the model with the smallest flexorder value from nextVisibleColumnModels
+    return nextVisibleColumnModels.reduce((smallestFlexOrderModel, currentModel) => {
+      if (smallestFlexOrderModel.flexOrder > currentModel.flexOrder) {
+        return currentModel;
+      }
+      return smallestFlexOrderModel;
+    });
+  }
+
+  get previousVisibleColumnModel(): ColumnOrderModelService {
+    // collect all columns that appears before this column
+    // that means collect all models with smaller flex order value
+    const previousVisibleColumnModels = this.columnOrderCoordinatorService.orderModels.filter(
+      model => model.flexOrder < this.flexOrder && !model.isHidden
+    );
+
+    // current column could be the one that visually appears at the first;
+    if (previousVisibleColumnModels.length === 0) {
+      return;
+    }
+
+    // to find the previous immediate visible columns model
+    // find the model with the largest flexorder value from previousVisibleColumnModels
+    return previousVisibleColumnModels.reduce((largestFlexOrderModel, currentModel) => {
+      if (largestFlexOrderModel.flexOrder < currentModel.flexOrder) {
+        return currentModel;
+      }
+      return largestFlexOrderModel;
+    });
+  }
+
+  get headerWidth(): number {
+    return this.headerEl ? this.domAdapter.clientRect(this.headerEl).width : 0;
```

Comment:
Making this one and the following two a lazy getter is risky for performance. Right now I can see it's called on `dragStart` so it's not that often, but you are re-computing 3 `clientRect`s for every single column at the same time. Depending on the browser and the size of the datagrid, this could get a bit slow.

Reply:
oh, I might have misunderstood your comment here. I could do a bit more than the current optimization I did.

Reply:
I get that you can't store a width forever. But in the same computation, aren't you re-computing every header width at least twice? Once for the drop tolerance of each side.

Reply:
Once we compute the width of a header, we recycle that for both sides' droppable.